### PR TITLE
[Enhancement] optimize mv rewrite plan performance

### LINKED ISSUE #18995: mv rewrite rules are time-consuming
## Enhancement
now mv rewrite rules are not that efficient. the following is time cost for ssb test with mv rewrite
| query / plan cost(ms) |  with mv   |    without mv |
| --- | --- | --- |
Q01 |54|14
Q02 |55|15
Q03 | 52|14
Q04 | 92|19
Q05 | 80|14
Q06 | 89|14
Q07 | 94|15
Q08 | 94|15
Q09 | 95|15
Q10 | 93|14
Q11 | 154|32
Q12 | 141|23
Q13 | 140|23

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #18995

## Problem Summary(Required) ：

optimize mv rewrite plan performance, including three points:
1. reduce mv rewrite rule application times by add matchedGroups
2. reduce view delta application times by only applying view delta rewrite on intersecting tables between mv and query, which can reduce memo space and rule application times
3. optimize mv rewrite rule performance, including removing extra log, reduce function call times and so on

| query / plan cost(ms) |  with mv   |    without mv | optimized with mv |
| --- | --- | --- | --- |
Q01 |54|14|13
Q02 |55|15|11
Q03 | 52|14|14
Q04 | 92|19|17
Q05 | 80|14|16
Q06 | 89|14|16
Q07 | 94|15|18
Q08 | 94|15|18
Q09 | 95|15|18
Q10 | 93|14|15
Q11 | 154|32|19
Q12 | 141|23|20
Q13 | 140|23|20

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.starrocks.sql.optimizer.base.ColumnRefFactory;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+
+import java.util.List;
+
+public class PlanContext {
+    private final OptExpression plan;
+    private final List<ColumnRefOperator> outputColumns;
+    private final ColumnRefFactory refFactory;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java:26]: move it to optimizerContext, and only transmit plan to mv？

Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java:26]: move to optimizerContext, and only transmit `plan` to mv？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java [hunk 2]
```diff
@@ -43,6 +48,7 @@ public class OptimizerContext {
     private OptimizerTraceInfo traceInfo;
     private OptimizerConfig optimizerConfig;
     private List<MaterializationContext> candidateMvs;
+    private Map<List<Table>, List<MaterializedView>> usedMvs;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java:51]: why don't add in `MaterializationContext`?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java [hunk 3]
```diff
@@ -141,4 +149,17 @@ public List<MaterializationContext> getCandidateMvs() {
...
     }
+
+    public void updateUsedMv(List<Table> tables, MaterializedView mv) {
+        List<Table> sortedTables = tables.stream()
+                .sorted((left, right) -> left.getName().compareTo(right.getName())).collect(Collectors.toList());
+        List<MaterializedView> mvs = usedMvs.computeIfAbsent(sortedTables, key -> Lists.newArrayList());
+        mvs.add(mv);
+    }
+
+    public List<MaterializedView> getUsedMvs(List<Table> tables) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java:160]: It looks like not like usedMvs, but like 'searchedMvs'

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java (hunk 2 above)
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java:51]: add some comment to desc key and value

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java:26]: As it's only used in `MaterializedView `, move it into `MaterializedView `as a inner class?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 4]
```diff
@@ -225,6 +228,7 @@ public List<OptExpression> rewrite(ReplaceColumnRefRewriter queryColumnRefRewrit
             if (rewrittenExpression == null) {
                 continue;
             }
+            materializationContext.getOptimizerContext().updateUsedMv(queryTables, materializationContext.getMv());
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:231]: no need update here as `updateUsedMv ` is called at `BaseMaterializedViewRewriteRule `

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java [hunk 5]
```diff
@@ -124,4 +135,20 @@ public List<Table> getBaseTables() {
     public boolean hasMultiTables() {
         return baseTables != null && baseTables.size() > 1;
     }
+
+    public Set<ColumnRefOperator> getOriginQueryColumns() {
+        return originQueryColumns;
+    }
+
+    public List<Table> getCommonTables() {
+        return commonTables;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java:144]: `commonTablesWIthQuery`: add some docs?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java (hunk 1 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java:26]: it is only used by Materialized view

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java (hunk 2 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java:51]: > why don't add it in `MaterializationContext`?

it is a context of query, not for mv.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java (hunk 4 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:231]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java (hunk 1 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/PlanContext.java:26]: mv PlanContext into MaterializedView as inner class

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java (hunk 2 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java]: mv it into MaterializedContext

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java (hunk 3 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/OptimizerContext.java]: move it into MaterializationContext, and change it to matchedTableLists

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java [hunk 6]
```diff
@@ -48,6 +50,19 @@ public boolean check(OptExpression input, OptimizerContext context) {
 
     @Override
     public List<OptExpression> transform(OptExpression queryExpression, OptimizerContext context) {
+        List<Table> queryTables = MvUtils.getAllTables(queryExpression);
+        List<MaterializedView> usedMvs = context.getUsedMvs(queryTables);
+        List<MaterializationContext> candidateContexts;
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRewriter.java [hunk 7]
```diff
@@ -121,23 +120,12 @@ public ScalarOperator visitVariableReference(ColumnRefOperator columnRef, Void c
                     return null;
                 }
                 Integer targetRelationId = srcToDstRelationIdMapping.get(srcRelationId);
-                List<ColumnRefOperator> relationColumns = dstRelationIdToColumns.get(targetRelationId);
+                Map<String, ColumnRefOperator> relationColumns = dstRelationIdToColumns.get(targetRelationId);
                 if (relationColumns == null) {
                     LOG.warn("no columns for relation id:{}", targetRelationId);
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRewriter.java]: done

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 8]
```diff
@@ -110,7 +110,25 @@ public boolean isValidPlan(OptExpression expression) {
...
+        // Only care MatchMode.COMPLETE and VIEW_DELTA here, QUERY_DELTA also can be supported
+        // because optimizer will match MV's pattern which is subset of query opt tree
+        // from top-down iteration.
+        if (matchMode == MatchMode.COMPLETE) {
+            if (!isJoinMatch(queryExpression, mvExpression, queryTables, mvTables)) {
+                return false;
+            }
+        } else if (matchMode == MatchMode.VIEW_DELTA) {
+            // only consider query with most common tables to optimize performance
+            if (!queryTables.containsAll(materializationContext.getCommonTables())) {
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java]: renamed to intersectingTables

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java [hunk 9]
```diff
@@ -124,4 +135,20 @@ public List<Table> getBaseTables() {
...
+    }
+
+    public List<Table> getQueryTables() {
```
[the remark this answers is not in the document]
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java]: add MvRewriteContext to pass argument for rewriter

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java [hunk 10]
```diff
@@ -15,50 +15,67 @@
...
+    private final List<Table> baseTables;
+
+    private final Set<ColumnRefOperator> originQueryColumns;
 
-    private List<Table> baseTables;
+    // tables both in query and mv
+    private final List<Table> intersectingTables;
 
-    public MaterializationContext(MaterializedView mv,
+    private List<List<Table>> matchedTableLists;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java:54]: List<Set<Table>> matchedTableLists? and plz add some  comment for this

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 11]
```diff
@@ -110,7 +113,25 @@ public boolean isValidPlan(OptExpression expression) {
...
+        // because optimizer will match MV's pattern which is subset of query opt tree
+        // from top-down iteration.
+        if (matchMode == MatchMode.COMPLETE) {
+            if (!isJoinMatch(queryExpression, mvExpression, queryTables, mvTables)) {
+                return false;
+            }
+        } else if (matchMode == MatchMode.VIEW_DELTA) {
+            // only consider query with most common tables to optimize performance
+            if (!queryTables.containsAll(materializationContext.getIntersectingTables())) {
+                return false;
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:129]: Will it actually happen？

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 12]
```diff
@@ -149,35 +171,31 @@ public List<OptExpression> rewrite(ReplaceColumnRefRewriter queryColumnRefRewrit
             return Lists.newArrayList();
         }
         ScalarOperator mvPredicate = MvUtils.rewriteOptExprCompoundPredicate(mvExpression, mvColumnRefRewriter);
+
         if (!ConstantOperator.TRUE.equals(mvPartitionPredicate)) {
+            // add latest partition predicate to mv predicate
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:176]: It looks like not suitable for comment here

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java (hunk 11 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:129]: yes

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java (hunk 12 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:176]: fixed

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java (hunk 10 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/MaterializationContext.java:54]: use groupid to replace List<Table>

**Review by Seaven — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by ABingHuang** [comment]
run starrocks_admit_test

**Comment by ABingHuang** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 148 / 169 (87.57%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|13|19|68.42%|[306, 307, 308, 309, 310, 311]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|36|48|75.00%|[733, 742, 752, 753, 760, 801, 802, 803, 804, 805, 806, 807]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializedViewOptimizer.java|6|7|85.71%|[38]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewritePreprocessor.java|17|19|89.47%|[91, 113]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|30|30|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/ColumnRewriter.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MaterializationContext.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/MvRewriteContext.java|12|12|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/rule/BaseMaterializedViewRewriteRule.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/ColumnRefFactory.java|1|1|100.00%|[]|

**Pull request merged**